[inductor] Fix unbacked floats being downcast to fp32 in Triton kernels

mlazos opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #170532
* #170531
* #170393
* __->__ #170289

Float scalars (both Python floats and unbacked float symbols from .item())
were being hardcoded to fp32 in Triton kernel signatures, causing precision
loss when the original value was fp64.

Fix by using fp64 for float scalars since:
- Python floats are natively fp64
- Unbacked floats from .item() on fp64 tensors should preserve precision

Add a numerics test that verifies bitwise equality between Python scalar
and tensor scalar paths for torch._foreach_addcmul with precision-sensitive
values like pi, 1/3, and the famous 0.1+0.2 issue.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

mlazos committed:
[inductor] Fix unbacked floats being downcast to fp32 in Triton kernels

Float scalars (both Python floats and unbacked float symbols from .item())
were being hardcoded to fp32 in Triton kernel signatures, causing precision
loss when the original value was fp64.

Fix by using fp64 for float scalars since:
- Python floats are natively fp64
- Unbacked floats from .item() on fp64 tensors should preserve precision

Add a numerics test that verifies bitwise equality between Python scalar
and tensor scalar paths for torch._foreach_addcmul with precision-sensitive
values like pi, 1/3, and the famous 0.1+0.2 issue.

[ghstack-poisoned]

mlazos committed:
Update on "[inductor] Fix unbacked floats being downcast to fp32 in Triton kernels"


Float scalars (both Python floats and unbacked float symbols from .item())
were being hardcoded to fp32 in Triton kernel signatures, causing precision
loss when the original value was fp64.

Fix by using fp64 for float scalars since:
- Python floats are natively fp64
- Unbacked floats from .item() on fp64 tensors should preserve precision

Add a numerics test that verifies bitwise equality between Python scalar
and tensor scalar paths for torch._foreach_addcmul with precision-sensitive
values like pi, 1/3, and the famous 0.1+0.2 issue.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

mlazos committed:
Update on "[inductor] Fix unbacked floats being downcast to fp32 in Triton kernels"


Float scalars (both Python floats and unbacked float symbols from .item())
were being hardcoded to fp32 in Triton kernel signatures, causing precision
loss when the original value was fp64.

Fix by using fp64 for float scalars since:
- Python floats are natively fp64
- Unbacked floats from .item() on fp64 tensors should preserve precision

Add a numerics test that verifies bitwise equality between Python scalar
and tensor scalar paths for torch._foreach_addcmul with precision-sensitive
values like pi, 1/3, and the famous 0.1+0.2 issue.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

eellison reviewed (approved):
Would you mind adding a test in https://github.com/pytorch/pytorch/blob/main/test/inductor/test_op_dtype_prop.py to show we handle this correctly there as well?

mlazos committed:
Update on "[inductor] Fix unbacked floats being downcast to fp32 in Triton kernels"


Float scalars (both Python floats and unbacked float symbols from .item())
were being hardcoded to fp32 in Triton kernel signatures, causing precision
loss when the original value was fp64.

Fix by using fp64 for float scalars since:
- Python floats are natively fp64
- Unbacked floats from .item() on fp64 tensors should preserve precision

Add a numerics test that verifies bitwise equality between Python scalar
and tensor scalar paths for torch._foreach_addcmul with precision-sensitive
values like pi, 1/3, and the famous 0.1+0.2 issue.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

mlazos committed:
Update on "[inductor] Fix unbacked floats being downcast to fp32 in Triton kernels"


Float scalars (both Python floats and unbacked float symbols from .item())
were being hardcoded to fp32 in Triton kernel signatures, causing precision
loss when the original value was fp64.

Fix by using fp64 for float scalars since:
- Python floats are natively fp64
- Unbacked floats from .item() on fp64 tensors should preserve precision

Add a numerics test that verifies bitwise equality between Python scalar
and tensor scalar paths for torch._foreach_addcmul with precision-sensitive
values like pi, 1/3, and the famous 0.1+0.2 issue.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

mlazos committed:
Update on "[inductor] Fix unbacked floats being downcast to fp32 in Triton kernels"


Float scalars (both Python floats and unbacked float symbols from .item())
were being hardcoded to fp32 in Triton kernel signatures, causing precision
loss when the original value was fp64.

Fix by using fp64 for float scalars since:
- Python floats are natively fp64
- Unbacked floats from .item() on fp64 tensors should preserve precision

Add a numerics test that verifies bitwise equality between Python scalar
and tensor scalar paths for torch._foreach_addcmul with precision-sensitive
values like pi, 1/3, and the famous 0.1+0.2 issue.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

mlazos committed:
Update on "[inductor] Fix unbacked floats being downcast to fp32 in Triton kernels"


Float scalars (both Python floats and unbacked float symbols from .item())
were being hardcoded to fp32 in Triton kernel signatures, causing precision
loss when the original value was fp64.

Fix by using fp64 for float scalars since:
- Python floats are natively fp64
- Unbacked floats from .item() on fp64 tensors should preserve precision

Add a numerics test that verifies bitwise equality between Python scalar
and tensor scalar paths for torch._foreach_addcmul with precision-sensitive
values like pi, 1/3, and the famous 0.1+0.2 issue.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

mlazos commented on torch/_inductor/codegen/triton_utils.py:
```diff
@@ -71,11 +71,13 @@ def signature_of(arg: KernelArgType, *, size_dtype: Optional[str]) -> str:
         elif isinstance(arg.expr, (float, sympy.Float)):
-            return "fp32"
+            # Python floats are natively fp64, so use fp64 to preserve precision
+            return "fp64"
```
yeah my hypothesis is correct, here's the repro:

https://gist.github.com/mlazos/240cdf885938bc1dcbcbe300688fff66

ngimel replied:
Discussed offline, @mlazos is correct, signature should be fp64, and then we will downcast prior to tensor ops

mlazos committed:
Update on "[inductor] Fix unbacked floats being downcast to fp32 in Triton kernels"


Float scalars (both Python floats and unbacked float symbols from .item())
were being hardcoded to fp32 in Triton kernel signatures, causing precision
loss when the original value was fp64.

Fix by using fp64 for float scalars since:
- Python floats are natively fp64
- Unbacked floats from .item() on fp64 tensors should preserve precision

Add a numerics test that verifies bitwise equality between Python scalar
and tensor scalar paths for torch._foreach_addcmul with precision-sensitive
values like pi, 1/3, and the famous 0.1+0.2 issue.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

mlazos committed:
Update on "[inductor] Fix unbacked floats being downcast to fp32 in Triton kernels"


Float scalars (both Python floats and unbacked float symbols from .item())
were being hardcoded to fp32 in Triton kernel signatures, causing precision
loss when the original value was fp64.

Fix by using fp64 for float scalars since:
- Python floats are natively fp64
- Unbacked floats from .item() on fp64 tensors should preserve precision

Add a numerics test that verifies bitwise equality between Python scalar
and tensor scalar paths for torch._foreach_addcmul with precision-sensitive
values like pi, 1/3, and the famous 0.1+0.2 issue.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

mlazos committed:
Update on "[inductor] Fix unbacked floats being downcast to fp32 in Triton kernels"


Float scalars (both Python floats and unbacked float symbols from .item())
were being hardcoded to fp32 in Triton kernel signatures, causing precision
loss when the original value was fp64.

Fix by using fp64 for float scalars since:
- Python floats are natively fp64
- Unbacked floats from .item() on fp64 tensors should preserve precision

Add a numerics test that verifies bitwise equality between Python scalar
and tensor scalar paths for torch._foreach_addcmul with precision-sensitive
values like pi, 1/3, and the famous 0.1+0.2 issue.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

mlazos committed:
Update on "[inductor] Fix unbacked floats being downcast to fp32 in Triton kernels"


Float scalars (both Python floats and unbacked float symbols from .item())
were being hardcoded to fp32 in Triton kernel signatures, causing precision
loss when the original value was fp64.

Fix by using fp64 for float scalars since:
- Python floats are natively fp64
- Unbacked floats from .item() on fp64 tensors should preserve precision

Add a numerics test that verifies bitwise equality between Python scalar
and tensor scalar paths for torch._foreach_addcmul with precision-sensitive
values like pi, 1/3, and the famous 0.1+0.2 issue.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

mlazos committed:
Update on "[inductor] Fix unbacked floats being downcast to fp32 in Triton kernels"


Float scalars (both Python floats and unbacked float symbols from .item())
were being hardcoded to fp32 in Triton kernel signatures, causing precision
loss when the original value was fp64.

Fix by using fp64 for float scalars since:
- Python floats are natively fp64
- Unbacked floats from .item() on fp64 tensors should preserve precision

Add a numerics test that verifies bitwise equality between Python scalar
and tensor scalar paths for torch._foreach_addcmul with precision-sensitive
values like pi, 1/3, and the famous 0.1+0.2 issue.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]
